src/islenska/basics.py

```diff
@@ -118,6 +123,138 @@
 CASES = ("NF", "ÞF", "ÞGF", "EF")
 CASES_LATIN = tuple(case.encode("latin-1") for case in CASES)
 
+# Symbols that make up a mark,
+# mapped to the format str argument for the mark form
+_MARK_ATOMS = ALL_CASES.union(
+    ALL_GENDERS,
+    ALL_NUMBERS,
+    (x.casefold() for x in ALL_BIN_PERSONS),
+    (x.casefold() for x in ALL_BIN_DEGREES),
+    (x.casefold() for x in ALL_BIN_TENSES),
+    (x.casefold() for x in ALL_BIN_VOICES),
+    (x.casefold() for x in ALL_BIN_MOODS),
+    (
+        "obeygjanlegt",
+        "serst",
+        "nogr",
+        "gr",
+        "það",
+        "st",
+        "lhþt",
+        "op",
+        "sp",
+        "sagnb",
+        "2",
+        "3",
+    ),
+)
+
+
+class MarkOrder:
+    """
+    Class used for ordering inflections and ensuring
+    an inflection exists for a specific word category.
+    """
+    # Singleton mark order dict
+    _order: Optional[Dict[str, Tuple[str, ...]]] = None
+
+    @classmethod
+    def _read_order_from_csv(cls):
+        """
+        Read ordering of inflections for
+        each word category from csv file.
+        """
+        p = Path(__file__).parent.resolve() / "resources" / "mark_order.csv"
+        assert p.exists() and p.is_file(), f"mark_order.csv file not found at {str(p)}"
+
+        o: Dict[str, List[str]] = {}
+        for line in p.read_text().splitlines():
+            [ordfl, mark] = line.split(";")
+            if ordfl not in o:
+                o[ordfl] = []
+            o[ordfl].append(mark)
+        # Create MarkOrder._order
+        cls._order = {k: tuple(v) for k, v in o.items()}
+
+    @classmethod
+    def index(cls, cat: str, m: str) -> int:
+        """
+        Find sorting index of a mark/inflection
+        for a given category.
+        """
+        if cls._order is None:
+            cls._read_order_from_csv()
+            assert cls._order is not None
+        # Added in order to sort 2/3 variant
+        # forms after 1 variant (normal form)
+        x = 0
+        if m.endswith("2"):
+            x = len(cls._order[cat])
+        elif m.endswith("3"):
+            x = 2 * len(cls._order[cat])
+        return cls._order[cat].index(m.rstrip("23")) + x
+
+    @classmethod
+    def valid_mark(cls, cat: str, m: str) -> bool:
+        """
+        Returns True if m is a valid mark/inflection specifier
+        for the given word category, False otherwise.
+        """
+        if cls._order is None:
+            cls._read_order_from_csv()
+            assert cls._order is not None
+        return m.rstrip("23") in cls._order[cat]
+
+
+def mark_to_set(mark: Union[str, Iterable[str]]) -> Set[str]:
+    """
+    Transform mark string into set of
+    inflection category specifiers.
+    """
+    if isinstance(mark, str):
+        mark = set(mark.split("-"))
+
+    atom_set: Set[str] = set()
+    for at in mark:
+        if "-" in at:
```

Comment:
Will this happen, i.e. that components of mark have hyphens? If mark was originally a string, it is split on hyphens, so it can never contain hyphens.

Reply:
In cases where the user could specify e.g. something like `to_inflection=("MM-VH", "ÞT")`, then the `MM-VH` atom would need to be split again. This just made it easier to call the function directly without having to break up each item on `-` when the user inputs an `Iterable` of inflection specifiers